### PR TITLE
Added support to store the status of export data snapshot for non-dbzm case in a file

### PR DESCRIPTION
**Pull request by priyanshi-yb**
Introduced a package as `jsonfile.go` which will be used to work with json files as read, write, update, etc...
Added a file in `metainfo` as `export_snapshot_status.json` (can't keep this in `data` dir as `pg_dump` needs an empty `data` directory to start with)  which is implemented using jsonfile to store export data status for a table for non-dbzm case,  and reading that in `export data status`/`get data-migration-report`. 
Oracle success run - https://jenkins.dev.yugabyte.com/job/users/job/yb-voyager-testing/job/yb-voyager-testing-pipeline/1750/

### CODE REVIEW

File: yb-voyager/src/srcdb/status.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+func prettifyExportSnapshotStatus(status *ExportSnapshotStatus) string {
+	jsonBytes, err := json.Marshal(status)
+	if err != nil {
+		panic(err)
+	}
+	statusJsonString := string(jsonBytes)
+	finalStatus := utils.PrettifyJsonString(statusJsonString)
+	return finalStatus
+}
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/status.go:104]: You can get rid of this by directly using `json.MarshalIndent(status, "", "    ")`

File: yb-voyager/cmd/exportDataStatus.go [hunk 2]
```diff
@@ -54,6 +57,15 @@ func initializeExportTableMetadata(tableList []*sqlname.SourceName) {
 		tablesProgressMetadata[key].Status = 0
 		tablesProgressMetadata[key].FileOffsetToContinue = int64(0)
 		tablesProgressMetadata[key].TableName = tableName
+		exportSnapshotStatus.Tables[key] = &srcdb.TableExportStatus{
+			TableName:                tableName.Qualified.MinQuoted,
+			FileName:                 "",
+			ExportedRowCountSnapshot: int64(0),
+		}
+	}
+	err := exportSnapshotStatus.Create()
```
Comment (amit-yb) [inline yb-voyager/cmd/exportDataStatus.go:66]: Will this work if `export data` is called multiple times?

File: yb-voyager/src/srcdb/status.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+func NewExportSnapshotStatus(exportDir string) *ExportSnapshotStatus {
+	return &ExportSnapshotStatus{
+		Tables:   make(map[string]*TableExportStatus),
+		FilePath: filepath.Join(exportDir, "metainfo", "export_snapshot_status.json"),
+	}
+}
+
+func ReadExportStatus(statusFilePath string) (*ExportSnapshotStatus, error) {
+	file, err := os.Open(statusFilePath)
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/status.go:35]: Consider implementing a utility class called `JsonFile` as:
src/utils/jsonfile/jsonfile.go
```
package jsonfile

import (
	"encoding/json"
	"fmt"
	"os"
	"sync"

	"github.com/yugabyte/yb-voyager/yb-voyager/src/utils"
)

type JsonFile[T any] struct {
	sync.Mutex
	FilePath string
}

func NewJsonFile[T any](filePath string) *JsonFile[T] {
	return &JsonFile[T]{FilePath: filePath}
}

func (j *JsonFile[T]) Read() (*T, error) {
	j.Lock()
	defer j.Unlock()
	return j.read()
}

func (j *JsonFile[T]) read() (*T, error) {
	bs, err := os.ReadFile(j.FilePath)
	if err != nil {
		return nil, fmt.Errorf("read file %s: %w", j.FilePath, err)
	}
	if len(bs) == 0 {
		return nil, fmt.Errorf("file %s is empty", j.FilePath)
	}
	obj := new(T)
	err = json.Unmarshal(bs, obj)
	if err != nil {
		return nil, fmt.Errorf("unmarshal json: %w", err)
	}
	return obj, nil
}

func (j *JsonFile[T]) Update(fn func(*T)) error {
	j.Lock()
	defer j.Unlock()
	var obj *T
	var err error
	if utils.FileOrFolderExists(j.FilePath) {
		obj, err = j.read()
		if err != nil {
			return err
		}
	} else {
		obj = new(T)
	}
	fn(obj)
	return j.write(obj)
}

func (j *JsonFile[T]) write(obj *T) error {
	bs, err := json.MarshalIndent(obj, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal json: %w", err)
	}
	err = os.WriteFile(j.FilePath, bs, 0644)
	if err != nil {
		return fmt.Errorf("write file %s: %w", j.FilePath, err)
	}
	return nil
}

func (j *JsonFile[T]) Delete() error {
	j.Lock()
	defer j.Unlock()
	return os.Remove(j.FilePath)
}
```
And `jsonfile_test.go`
```
package jsonfile

import (
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestJsonFile(t *testing.T) {
	type Person struct {
		Name string
	}
	var person *Person
	jf := NewJsonFile[Person]("person.json")
	_ = jf.Delete()
	person, err := jf.Read()
	assert.ErrorIs(t, err, os.ErrNotExist)
	assert.Nil(t, person)
	err = jf.Update(func(p *Person) {
		p.Name = "John Doe"
	})
	assert.Nil(t, err)
	person, err = jf.Read()
	assert.Nil(t, err)
	assert.NotNil(t, person)
	assert.Equal(t, "John Doe", person.Name)
	err = jf.Update(func(p *Person) {
		p.Name = "John Smith"
	})
	assert.Nil(t, err)
	person, err = jf.Read()
	assert.Nil(t, err)
	assert.Equal(t, "John Smith", person.Name)
	err = jf.Delete()
	assert.Nil(t, err)
}
```
Once you have the `JsonFile` implementation:
- you can make the `ExportSnapshotStatus` a plain data object without any methods.
- you can directly define and use `exportSnapshotStatusFile JsonFile[ExportSnapshotStatus]`.

**Review by amit-yb — COMMENTED** [review]
Let's implement a generic `JsonFile` type implement the `export_snapshot_status.json` using it.

File: yb-voyager/cmd/exportDataStatus.go (hunk 2 above)
Comment (priyanshi-yb) [reply yb-voyager/cmd/exportDataStatus.go]: yeah, deleting it on start-clean so create will work on every run of export data which can happen with start-clean only.

**Review by amit-yb — APPROVED** [review]
LGTM

File: yb-voyager/src/srcdb/exportDataStatus.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package srcdb
```
[the remark this answers is not in the document]
Comment (priyanshi-yb) [reply yb-voyager/src/srcdb/exportDataStatus.go:1]: hmm, was also thinking the same. Will move it to `cmd` now.

**Pull request merged**